library/src/main/java/com/datatorrent/lib/join/AbstractJoinOperator.java

```diff
@@ -0,0 +1,343 @@
+    }
+    // Checks whether the strategy is outer join and set it to store
+    boolean isOuter = strategy.equals(JoinStrategy.LEFT_OUTER_JOIN) || strategy.equals(JoinStrategy.OUTER_JOIN);
+    store[0].isOuterJoin(isOuter);
+    isOuter = strategy.equals(JoinStrategy.RIGHT_OUTER_JOIN) || strategy.equals(JoinStrategy.OUTER_JOIN);
+    store[1].isOuterJoin(isOuter);
+    // Setup the stores
+    store[0].setup();
+    store[1].setup();
+
+    populateFields();
+  }
+
+  /**
+   * Create the event with the given tuple. If it successfully inserted it into the store
+   * then it does the join operation
+   *
+   * @param tuple Tuple to process
+   */
+  protected void processTuple(T tuple)
+  {
+    int idx = 0;
+    if (!isLeft) {
+      idx = 1;
+    }
+    TimeEvent t = createEvent(tuple);
+    if (store[idx].put(t)) {
+      join(t, isLeft);
+    }
+  }
+
+  private void populateFields()
+  {
+    populateIncludeFields();
+    populateKeyFields();
+    if (timeFieldStr != null) {
+      populateTimeFields();
+    }
+  }
+
+  /**
+   * Populate the fields from the includeFiledStr
+   */
+  private void populateIncludeFields()
+  {
+    includeFields = new String[2][];
+    String[] portFields = includeFieldStr.split(";");
+    for (int i = 0; i < portFields.length; i++) {
+      includeFields[i] = portFields[i].split(",");
+    }
+  }
+
+  /**
+   * Get the tuples from another store based on join constraint and key
+   *
+   * @param tuple  input
+   * @param isLeft whether the given tuple is from first port or not
+   */
+  private void join(TimeEvent tuple, Boolean isLeft)
+  {
+    // Get the valid tuples from the store based on key
+    // If the tuple is null means the join type is outer and return unmatched tuples from store.
+    Object value;
+    if (isLeft) {
+      if (tuple != null) {
+        value = store[1].getValidTuples(tuple);
+      } else {
+        value = store[1].getUnMatchedTuples();
+      }
+    } else {
+      if (tuple != null) {
+        value = store[0].getValidTuples(tuple);
+      } else {
+        value = store[0].getUnMatchedTuples();
+      }
+    }
+    // Join the input tuple with the joined tuples
+    if (value != null) {
+      ArrayList<TimeEvent> joinedValues = (ArrayList<TimeEvent>)value;
+      List<T> result = new ArrayList<>();
+      for (TimeEvent joinedValue : joinedValues) {
+        T output = createOutputTuple();
+        Object tupleValue = null;
+        if (tuple != null) {
+          tupleValue = tuple.getValue();
+        }
+        copyValue(output, tupleValue, isLeft);
+        copyValue(output, joinedValue.getValue(), !isLeft);
+        result.add(output);
+        joinedValue.setMatch(true);
+      }
+      if (tuple != null) {
+        tuple.setMatch(true);
+      }
+      if (result.size() != 0) {
+        outputPort.emit(result);
+      }
+    }
+  }
+
+  // Emit the unmatched tuples, if the strategy is outer join
+  @Override
```

Comment:
Here the case is, if the strategy is not INNER_JOIN. In this case, the tuples which are not applied the join operation in expiry time period, will apply the join operation after expiry .